KAFKA-3492: Secure quotas for authenticated users

rajinisivaram opened the pull request:
Implementation and tests for secure quotas at <user> and <user, client-id> levels as described in KIP-55. Also adds dynamic default quotas for <client-id>, <user> and <user-client-id>. For each client connection, the most specific quota matching the connection is used, with user quota taking precedence over client-id quota.

rajinisivaram committed:
KAFKA-3492: Secure quotas for authenticated users

rajinisivaram commented:
This PR has been updated to match the final version of KIP-55. Feedback appreciated. Thank you...

rajinisivaram commented:
@junrao @ijuma Would either of you have time to review this PR? Thank you...

junrao commented:
@rajinisivaram : Sorry for the delay. I will take a look next week.

rajinisivaram commented:
@junrao Would you be a able to review this in time to include in 0.10.1.0? Thank you...

junrao commented on core/src/main/scala/kafka/admin/AdminUtils.scala:
```diff
@@ -498,37 +515,41 @@ object AdminUtils extends Logging {
-  private def changeEntityConfig(zkUtils: ZkUtils, entityType: String, entityName: String, configs: Properties) {
+  private def changeEntityConfig(zkUtils: ZkUtils, rootEntityType: String, fullSanitizedEntityName: String, configs: Properties) {
+    val sanitizedEntityPath = rootEntityType + '/' + fullSanitizedEntityName
+    val entityConfigPath = getEntityConfigPath(rootEntityType, fullSanitizedEntityName)
```
It seems that sanitizedEntityPath and entityConfigPath are the same?

rajinisivaram replied:
`sanitizedEntityPath` is `entityType/entityName`. This is the value stored in the JSON for config updates. `entityConfigPath` is the full path in Zookeeper. It is `/config/entityType/entityName`.

junrao commented on core/src/main/scala/kafka/admin/AdminUtils.scala:
```diff
@@ -560,6 +581,23 @@ object AdminUtils extends Logging {
     zkUtils.getAllEntitiesWithConfig(entityType).map(entity => (entity, fetchEntityConfig(zkUtils, entityType, entity))).toMap
 
+  def fetchAllChildEntityConfigs(zkUtils: ZkUtils, rootEntityType: String, childEntityType: String): Map[String, Properties] = {
+    def entityPaths(zkUtils: ZkUtils, rootPath: Option[String]) = {
```
Could we explicitly specify the return type?

rajinisivaram replied:
Done.

junrao commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -21,27 +21,38 @@ import java.util.Properties
 import org.apache.kafka.common.security.JaasUtils
 import org.apache.kafka.common.utils.Utils
 
+import scala.collection.JavaConverters._
```
duplicate import

rajinisivaram replied:
Removed.

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -16,16 +16,20 @@
 import org.apache.kafka.common.utils.Time
 
+import scala.collection.JavaConversions._
+import scala.collection.JavaConverters._
```
redundant import

rajinisivaram replied:
Removed.

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -54,11 +59,84 @@ object ClientQuotaManagerConfig {
+
+object QuotaId {
+
+  val Separator = ':'
```
Will this cause issues in jmx name? It seems that : is used to separate types in jmx? Similarly, are %, <, > allowed in jmx name? Is < > allows in ZK path?

rajinisivaram replied:
The separator is used only in sensor name. Metrics store user and clientId as separate tags. To avoid confusion, I have moved the concatenation to sensor names alone.

I think % is valid in JMX, but just in case some tools don't like it, I have changed sanitize method to encode underscore and then replace % with underscores. So now an encoded user principal contains the same set of characters as clientIds, which should make it valid everywhere.
`<default>` is used in Zookeeper nodes for default configuration, but never in metrics (metrics use actual clientIds and users only). Zookeeper docs (https://zookeeper.apache.org/doc/r3.1.2/zookeeperProgrammers.html#ch_zkDataModel) say that this should be valid. I have tested it too and it seems fine.

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -54,11 +59,84 @@ object ClientQuotaManagerConfig {
+  /**
+   * Decodes sanitized user principal
+   */
+  def desanitize(sanitizedUser: String) = {
```
Could we explicitly specify the return type? Ditto in a few other places.

rajinisivaram replied:
Done.

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -54,11 +59,84 @@ object ClientQuotaManagerConfig {
+   * Decodes sanitized user principal
+   */
+  def desanitize(sanitizedUser: String) = {
+    URLDecoder.decode(sanitizedUser, StandardCharsets.UTF_8.name)
```
Do we need to handle \* and + specially here?

rajinisivaram replied:
URL decoder decodes all % encoded sequences, so special treatment is required for \* and +. Have added a unit test to verify this too.

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -257,31 +449,84 @@ class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
+            val metricConfigEntity = quotaEntity(sanitizedUser.getOrElse(""), clientId.getOrElse(""))
+            val newQuota = metricConfigEntity.quota
+            logger.info(s"Sensor for ${userInfo}${clientIdInfo} already exists. Changing quota to ${newQuota.bound()} in MetricConfig")
+            allMetrics.get(quotaMetricName).config(getQuotaMetricConfig(newQuota))
```
This is an existing corner case issue. It's possible for a metric to get expired between the time the check in line 490 and here. So, it's safer to save the result of allMetrics.get(quotaMetricName) to a local val and only update the config if it's not null.

rajinisivaram replied:
Done.

junrao commented on core/src/main/scala/kafka/server/DynamicConfigManager.scala:
```diff
@@ -55,8 +62,10 @@ object ConfigType {
+ * The sequential znode contains data in this format: {"version" : 1, "entity_type":"topic/client", "entity_name" : "topic_name/client_id"}
  * This is just a notification--the actual config change is stored only once under the /config/entityType/entityName path.
+ * Version 2 of notifications has the format: {"version" : 2, "entity_path":"entity_type/entity_name"}
+ * Multiple entities may be specified as a hiearachical path (eg. users/<user>/clients/<clientId>).
```
typo hiearachical

rajinisivaram replied:
Fixed.

junrao commented on core/src/main/scala/kafka/server/DynamicConfigManager.scala:
```diff
@@ -79,6 +88,22 @@ class DynamicConfigManager(private val zkUtils: ZkUtils,
                            private val time: Time = SystemTime) extends Logging {
   private var lastExecutedChange = -1L
 
+  // Apply all existing client/user configs to the ClientIdConfigHandler/UserConfigHandler to bootstrap the overrides
```
The subscription to ZK change notification path was made in ZkNodeChangeNotificationListener. So, in order not to miss any changes, we should apply the bootstrap logic after ZkNodeChangeNotificationListener is initialized.

rajinisivaram replied:
Moved the code to `startup` after the listener is initialized.

junrao commented on core/src/main/scala/kafka/server/DynamicConfigManager.scala:
```diff
@@ -87,30 +112,60 @@ class DynamicConfigManager(private val zkUtils: ZkUtils,
       }
     }
+
+    private def processEntityConfigChangeVersion1(json: String, map: Map[String, Any]) = {
```
Does this method need to return any value?

rajinisivaram replied:
No, removed unnecessary =.

junrao commented on core/src/main/scala/kafka/server/DynamicConfigManager.scala:
```diff
@@ -87,30 +112,60 @@ class DynamicConfigManager(private val zkUtils: ZkUtils,
+
+    }
+
+    private def processEntityConfigChangeVersion2(json: String, map: Map[String, Any]) = {
```
Does this method need to return any value?

rajinisivaram replied:
No, removed unnecessary =.

junrao commented on core/src/test/scala/integration/kafka/api/BaseQuotaTest.scala:
```diff
@@ -0,0 +1,210 @@
+import java.util.Properties
+
+import kafka.consumer.SimpleConsumer
+import kafka.integration.KafkaServerTestHarness
```
unused import

rajinisivaram replied:
Removed.

junrao commented on core/src/test/scala/integration/kafka/api/ClientIdQuotaTest.scala:
```diff
@@ -0,0 +1,62 @@
+import java.util.Properties
+
+import kafka.admin.AdminUtils
+import kafka.server.{KafkaConfig, QuotaConfigOverride, ClientQuotaManagerConfig, QuotaId}
```
unused import ClientQuotaManagerConfig

rajinisivaram replied:
Removed.

junrao commented on core/src/test/scala/integration/kafka/api/UserClientIdQuotaTest.scala:
```diff
@@ -0,0 +1,66 @@
+
+import kafka.admin.AdminUtils
+
+import kafka.server.{KafkaConfig, ClientQuotaManagerConfig, ConfigEntityName, QuotaConfigOverride, QuotaId}
```
unused import ClientQuotaManagerConfig

rajinisivaram replied:
Removed.

junrao commented on core/src/test/scala/integration/kafka/api/UserQuotaTest.scala:
```diff
@@ -0,0 +1,61 @@
+import java.util.Properties
+
+import kafka.admin.AdminUtils
+import kafka.server.{KafkaConfig, ClientQuotaManagerConfig, ConfigEntityName, QuotaConfigOverride, QuotaId}
```
unused import ClientQuotaManagerConfig and QuotaConfigOverride

junrao commented:
@rajinisivaram : Sorry for the late review. I made a pass and left some comments. 

Also, could update the design doc on quota related stuff?

rajinisivaram commented:
@junrao Thank you very much for the review. I have addressed all your comments. I will update the design docs and document the config changes under KAFKA-4079.

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -54,11 +58,73 @@ object ClientQuotaManagerConfig {
+    for (i <- 0 until encoded.length) {
+      encoded.charAt(i) match {
+        case '*' => builder.append("%2A") // Metric ObjectName treats * as pattern
+        case '+' => builder.append("%20") // Space URL-encoded as +, replace with percent encoding
```
So URL encoder doesn't encode \* and + automatically?

rajinisivaram replied:
URLEncoder doesn't encode '*'. It does encode '+', but space is then encoded as '+'. So we replace '+' with %20 which is percent encoding for space.

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -209,19 +391,15 @@ class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
-                                                "Tracking average throttle-time per client",
-                                                "client-id",
-                                                clientId), new Avg())
+          throttleTimeSensor.add(throttleMetricName(quotaEntity), new Avg())
```
Now that quota is a bit more complicated, would it be useful to log quotaEntity in debug level every time a sensor is created for debugging purpose?

rajinisivaram replied:
Done.

junrao commented on core/src/test/scala/integration/kafka/api/BaseQuotaTest.scala:
```diff
@@ -0,0 +1,208 @@
+    consume(consumers.head, numRecords)
+    // The replica consumer should not be throttled also. Create a fetch request which will exceed the quota immediately
+    val request = new FetchRequestBuilder().addFetch(topic1, 0, 0, 1024*1024).replicaId(followerNode.config.brokerId).build()
+    replicaConsumers.head.fetch(request)
```
This is an existing issue. Not sure how we test that replica consumer is not throttled here. If it doesn't, perhaps we can remove replicaConsumers?

rajinisivaram replied:
Yes, I was wondering about that too. Removed.

junrao commented on core/src/test/scala/integration/kafka/api/BaseQuotaTest.scala:
```diff
@@ -0,0 +1,208 @@
+    consume(consumers.head, numRecords)
+    // The replica consumer should not be throttled also. Create a fetch request which will exceed the quota immediately
+    val request = new FetchRequestBuilder().addFetch(topic1, 0, 0, 1024*1024).replicaId(followerNode.config.brokerId).build()
+    replicaConsumers.head.fetch(request)
```
Ditto as the above.

rajinisivaram replied:
Removed.

junrao commented on core/src/test/scala/integration/kafka/api/ClientIdQuotaTest.scala:
```diff
@@ -0,0 +1,62 @@
+class ClientIdQuotaTest extends BaseQuotaTest {
+
+  override def generateConfigs() = {
+    FixedPortTestUtils.createBrokerConfigs(serverCount,
```
Is there a reason to use FixedPortTestUtils, which is discouraged since it can cause port conflicts?

rajinisivaram replied:
I wasn't sure why it was there, so I left it as is for the client-id quota test. I have tested without and it works fine, so I have removed that.

junrao commented on core/src/test/scala/integration/kafka/api/UserQuotaTest.scala:
```diff
@@ -0,0 +1,61 @@
+    updateQuotaOverride(consumerClientId, emptyProps)
+  }
+
+  private def updateQuotaOverride(clientId: String, properties: Properties) {
```
It seems that clientId is not used?

rajinisivaram replied:
Removed unused parameter.

junrao commented on core/src/test/scala/unit/kafka/admin/ConfigCommandTest.scala:
```diff
@@ -65,4 +69,183 @@ class ConfigCommandTest extends ZooKeeperTestHarness with Logging {
+
+    def checkEntities(opts: Array[String], expectedFetches: Map[String, Seq[String]], expectedEntityNames: Seq[String]) {
+      val entity = ConfigCommand.parseEntity(new ConfigCommandOptions(opts :+ "--describe"))
+      val root = entity.root.entityType + "/" + entity.fullSanitizedName
```
root is unused?

rajinisivaram replied:
Removed.

junrao commented on core/src/test/scala/unit/kafka/server/ClientQuotaManagerTest.scala:
```diff
@@ -38,43 +39,171 @@ class ClientQuotaManagerTest {
+      assertEquals("Should return the newly overridden value (4000)", new Quota(4000, true), clientMetrics.quota(client1.user, client1.clientId))
+
+      throttleTimeMs = clientMetrics.recordAndMaybeThrottle(client1.user, client1.clientId, 0, this.callback)
+      assertEquals(s"throttleTimeMs should be 0. was $throttleTimeMs", 0, throttleTimeMs)
```
Should we record 0 here?

rajinisivaram replied:
Changed value to > 0, less than new default, checks that client is not throttled.

junrao commented on core/src/test/scala/unit/kafka/server/ClientQuotaManagerTest.scala:
```diff
@@ -38,43 +39,171 @@ class ClientQuotaManagerTest {
+  }
+
+  @Test
+  def testUserQuotaParsingWithDefaultClientIdQuota() {
```
Could you add a comment to explain how this is different from testUserQuotaParsing()?

rajinisivaram replied:
Done.

junrao commented:
@rajinisivaram : Could you test your branch on our system tests? You can use https://jenkins.confluent.io/job/system-test-kafka-branch-builder/. 

Also, there is an existing system test on client quota (tests/kafkatest/tests/client/quota_test.py). Should we expand that to test the new options?

rajinisivaram commented:
@junrao Thanks again for the review. Have addressed the comments.

I have scheduled a system test run, will add a link here when the test starts. Also planning to add system tests for the new configs under KAFKA-4055.

rajinisivaram commented:
System test results for this branch are here: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/534/console. There were two unrelated failures which look like existing timing issues.

junrao commented:
@rajinisivaram : Thanks for the update. There is another PR (https://github.com/apache/kafka/pull/1776) for KIP-73 that has gone through a few rounds of reviews and is pretty close to be committed. It will conflict with your PR. Since that PR is larger, I am thinking of committing that PR first (should be today or early tomorrow). Do you think you can rebase after that PR is merged? Also, do you think you can patch the client quota system test within a week or two?

rajinisivaram commented:
@junrao Yes, I can rebase tomorrow after the other one is merged. I submitted the quota system test updates under KAFKA-4055 this morning - PR #1860. Thanks.

junrao commented:
@rajinisivaram : Thanks. KIP-73 is committed now.

rajinisivaram commented:
@junrao Thank you, I have rebased the PR.

junrao commented on core/src/test/scala/unit/kafka/admin/TestAdminUtils.scala:
```diff
@@ -23,5 +23,6 @@ class TestAdminUtils extends AdminUtilities {
   override def changeBrokerConfig(zkUtils: ZkUtils, brokerIds: Seq[Int], configs: Properties): Unit = {}
   override def fetchEntityConfig(zkUtils: ZkUtils, entityType: String, entityName: String): Properties = {new Properties}
   override def changeClientIdConfig(zkUtils: ZkUtils, clientId: String, configs: Properties): Unit = {}
+  def changeUserOrUserClientIdConfig(zkUtils: ZkUtils, sanitizedEntityName: String, configs: Properties): Unit = {}
```
Should we add override ?

rajinisivaram replied:
Done.

junrao commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -150,6 +159,115 @@ object ConfigCommand {
+      // Describe option examples:
+      //   Single entity name:
+      //     --entity-type topics --entity-name topic1 (topic1)
+      //     --entity-type users --entity-name user1 --entity-type clients --entity-name client2 (<user1, client2>)
```
This doesn't seem to be a single entity as the comment in line 189 suggests? Ditto for line 193.

rajinisivaram replied:
Updated comment.

junrao commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -195,14 +317,23 @@ object ConfigCommand {
       CommandLineUtils.checkInvalidArgs(parser, options, describeOpt, Set(alterOpt, addConfig, deleteConfig))
       if(options.has(alterOpt)) {
-        require(options.has(entityName), "--entity-name must be specified with --alter")
+        if(!options.has(entityName) && !options.has(entityDefault))
```
entityDefault can only be specified for client/user for now. So we probably need to adjust the check a bit.

rajinisivaram replied:
Done.

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -53,11 +58,73 @@ object ClientQuotaManagerConfig {
+      encoded.charAt(i) match {
+        case '*' => builder.append("%2A") // Metric ObjectName treats * as pattern
+        case '+' => builder.append("%20") // Space URL-encoded as +, replace with percent encoding
+        case '_' => builder.append("%5F") // Encode underscore and replace % with underscore since some metrics tools might not handle %
```
Could you explain why we need to encode _?

rajinisivaram replied:
I was thinking that if there are tools that dont like %, it may be useful to replace % as well, and I picked underscore as the character to replace %. But having looked into it a bit more, I think this is unnecessary. % does appear elsewhere in metrics, so it should be safe to leave it. Have removed the encoding of _.

junrao commented on core/src/main/scala/kafka/server/DynamicConfigManager.scala:
```diff
@@ -122,5 +161,21 @@ class DynamicConfigManager(private val zkUtils: ZkUtils,
+          AdminUtils.fetchAllEntityConfigs(zkUtils, ConfigType.Client).foreach {
+            case (clientId, properties) => handler.processConfigChanges(clientId, properties)
+          }
+      case (ConfigType.User, handler) =>
```
Hmm, should we handle topic and broker config types too?

rajinisivaram replied:
Done.

junrao commented on core/src/test/scala/integration/kafka/api/BaseQuotaTest.scala:
```diff
@@ -0,0 +1,196 @@
+
+import java.util.Properties
+
+import kafka.consumer.SimpleConsumer
```
A few unused imports.

rajinisivaram replied:
Removed.

junrao commented:
1. I ran the following command and it's weird to see a user level only config in the output.

bin/kafka-configs.sh --zookeeper localhost:2181 --describe --entity-type users --entity-type clients
Configs for user-principal 'junrao' are producer_byte_rate=100
Configs for user-principal 'junrao', client-id 'client1' are producer_byte_rate=300
1. If I add a client + user default config and run the command again, it's weird to see an empty default user-principal in the output.

Configs for default user-principal are 
Configs for default user-principal, client-id 'client1' are producer_byte_rate=300
Configs for user-principal 'junrao' are producer_byte_rate=100
Configs for user-principal 'junrao', client-id 'client1' are producer_byte_rate=300

junrao commented:
@rajinisivaram : Thanks for the patch. I left a few minor comments. Once they are addressed, I can merge the PR in.

rajinisivaram committed:
KAFKA-3492: Address review comments

rajinisivaram commented:
@junrao Thank you for the review. I was thinking that it would be useful to have one kafka-configs describe command that described user as well as all the clients under it. But as you have pointed out, it is not very intuitive. So I have changed the command to list only <user, client> when both user and client types are specified. 

Have addressed all the other comments too. Thanks.

junrao commented on core/src/main/scala/kafka/server/DynamicConfigManager.scala:
```diff
@@ -122,5 +161,29 @@ class DynamicConfigManager(private val zkUtils: ZkUtils,
+          AdminUtils.fetchAllEntityConfigs(zkUtils, ConfigType.Topic).foreach {
+            case (topic, properties) => handler.processConfigChanges(topic, properties)
+          }
+      case _ =>
```
It seems that we can use the catch all case to handle all client, broker and topic?

rajinisivaram replied:
Done.

junrao commented:
Also, when I ran the following command,

bin/kafka-configs.sh --zookeeper localhost:2181 --describe --entity-type users

I got the following. 

Configs for default user-principal are 
Configs for user-principal 'junrao' are producer_byte_rate=100

Could we not show "Configs for default user-principal" if it's empty?

rajinisivaram committed:
KAFKA-3492: Address review comments

rajinisivaram commented:
@junrao Thanks again. Have fixed the --describe for users.

rajinisivaram committed:
KAFKA-3492: Improve log

junrao commented:
@rajinisivaram : Thanks a lot for the patch. LGTM
